### PR TITLE
PE 1717 Shared File Password Unlock

### PR DESCRIPTION
**Pull request by javdhu**
Adds an unlock screen to the shared file page if a fileKey is not present.

### CODE REVIEW

File: lib/l10n/app_en.arb [hunk 1]
```diff
@@ -983,6 +991,14 @@
   "@specifiedFileDoesNotExist": {
     "description": "The shared file the user is trying to access does not exist"
   },
+  "sharedFileInvalidFileKey": "The file key is invalid.",
+  "@sharedFileInvalidFileKey": {
+    "description": "The user entered an incorrect file key"
+  },
+  "sharedFileIsEnrypted": "This file is encrypted.",
+  "@sharedFileIsEnrypted": {
```
Comment (thiagocarvalhodev) [inline lib/l10n/app_en.arb:999]:
```suggestion
  "@sharedFileIsEncrypted":
```

Comment (thiagocarvalhodev) [inline lib/l10n/app_en.arb:999]:
```suggestion
  "@sharedFileIsEncrypted": {
```

File: lib/pages/shared_file/shared_file_page.dart [hunk 2]
```diff
@@ -36,6 +40,32 @@ class SharedFilePage extends StatelessWidget {
                     fit: BoxFit.contain,
                   ),
                   const SizedBox(height: 32),
+                  if (state is SharedFileIsPrivate) ...[
+                    Text(appLocalizationsOf(context).sharedFileIsEnrypted),
```
Comment (thiagocarvalhodev) [inline lib/pages/shared_file/shared_file_page.dart:44]:
```suggestion
                    Text(appLocalizationsOf(context).sharedFileIsEncrypted),
```

File: lib/pages/shared_file/shared_file_page.dart [hunk 3]
```diff
@@ -78,6 +108,6 @@ class SharedFilePage extends StatelessWidget {
 
   Widget _buildReturnToAppLink() => TextButton(
         onPressed: () => launch('https://ardrive.io/'),
-        child: Text('Learn more about ArDrive'),
+        child: const Text('Learn more about ArDrive'),
```
Comment (thiagocarvalhodev) [inline lib/pages/shared_file/shared_file_page.dart:111]: Do we have a ticket for apply internalization on this?

File: lib/blocs/shared_file/shared_file_cubit.dart [hunk 4]
```diff
@@ -18,16 +23,74 @@ class SharedFileCubit extends Cubit<SharedFileState> {
 
   final ArweaveService? _arweave;
 
-  SharedFileCubit({this.fileId, this.fileKey, ArweaveService? arweave})
+  SharedFileCubit({required this.fileId, this.fileKey, ArweaveService? arweave})
```
Comment (thiagocarvalhodev) [inline lib/blocs/shared_file/shared_file_cubit.dart:26]: Consider `arweave` a non-nullable variable and as a required parameter

File: lib/pages/shared_file/shared_file_page.dart (hunk 2 above)
Comment (jdaev) [reply lib/pages/shared_file/shared_file_page.dart]: [Done](https://github.com/ardriveapp/ardrive-web/pull/578/commits/7ce3ef522fd13cf024e89e00e2ef0a82e6cf6184)

File: lib/l10n/app_en.arb (hunk 1 above)
Comment (jdaev) [reply lib/l10n/app_en.arb]: [Corrected](https://github.com/ardriveapp/ardrive-web/pull/578/commits/7ce3ef522fd13cf024e89e00e2ef0a82e6cf6184)

File: lib/pages/app_router_delegate.dart [hunk 5]
```diff
@@ -97,7 +97,7 @@ class AppRouterDelegate extends RouterDelegate<AppRoutePath>
             shell = BlocProvider<SharedFileCubit>(
               key: ValueKey(sharedFileId),
               create: (_) => SharedFileCubit(
-                fileId: sharedFileId,
+                fileId: sharedFileId!,
```
Comment (karlprieb) [inline lib/pages/app_router_delegate.dart:100]: Why this change?

Comment (jdaev) [reply lib/pages/app_router_delegate.dart:100]: It was nullable inside the cubit before. Now that's required and we force unwrap here instead.
There's a null check above this so this is safe.

**Review by karlprieb — APPROVED** [review]

**Review by thiagocarvalhodev — DISMISSED** [review]

**Comment by thiagocarvalhodev** [comment]
I'm just missing error handling at: shared_file_cubit.dart.

For instance we have this function that if we get an error on `_arweave` the user will still seeing the load in progress.

```dart
void submit() async {
    form.markAllAsTouched();

    if (form.invalid) {
      return;
    }

    emit(SharedFileLoadInProgress());
    final String? fileKeyBase64 = form.control('fileKey').value;
    final fileKey = SecretKey(decodeBase64ToBytes(fileKeyBase64!));
    final file = await _arweave.getLatestFileEntityWithId(fileId, fileKey);

    if (file != null) {
      emit(SharedFileLoadSuccess(file: file, fileKey: fileKey));
    } else {
      emit(SharedFileNotFound());
    }
  }
```
We can do

``` dart
void submit() async {
    try{
      form.markAllAsTouched();

      if (form.invalid) {
        return;
      }

      emit(SharedFileLoadInProgress());
      final String? fileKeyBase64 = form.control('fileKey').value;
      final fileKey = SecretKey(decodeBase64ToBytes(fileKeyBase64!));
      final file = await _arweave.getLatestFileEntityWithId(fileId, fileKey);

      if (file != null) {
        emit(SharedFileLoadSuccess(file: file, fileKey: fileKey));
      } else {
        emit(SharedFileNotFound());
       }
    } catch(e) {
      emit(SharedFileError());
    }
  }
  ```

**Comment by jdaev** [comment]
> I'm just missing error handling at: shared_file_cubit.dart.
> 
> For instance we have this function that if we get an error on `_arweave` the user will still seeing the load in progress.
> 
> ```dart
> void submit() async {
>     form.markAllAsTouched();
> 
>     if (form.invalid) {
>       return;
>     }
> 
>     emit(SharedFileLoadInProgress());
>     final String? fileKeyBase64 = form.control('fileKey').value;
>     final fileKey = SecretKey(decodeBase64ToBytes(fileKeyBase64!));
>     final file = await _arweave.getLatestFileEntityWithId(fileId, fileKey);
> 
>     if (file != null) {
>       emit(SharedFileLoadSuccess(file: file, fileKey: fileKey));
>     } else {
>       emit(SharedFileNotFound());
>     }
>   }
> ```
> 
> We can do
> 
> ```dart
> void submit() async {
>     try{
>       form.markAllAsTouched();
> 
>       if (form.invalid) {
>         return;
>       }
> 
>       emit(SharedFileLoadInProgress());
>       final String? fileKeyBase64 = form.control('fileKey').value;
>       final fileKey = SecretKey(decodeBase64ToBytes(fileKeyBase64!));
>       final file = await _arweave.getLatestFileEntityWithId(fileId, fileKey);
> 
>       if (file != null) {
>         emit(SharedFileLoadSuccess(file: file, fileKey: fileKey));
>       } else {
>         emit(SharedFileNotFound());
>        }
>     } catch(e) {
>       emit(SharedFileError());
>     }
>   }
> ```

There is error handling inside getLatestFileEntityWithId. It will throw and return null.

**Comment by thiagocarvalhodev** [comment]
> There is error handling inside getLatestFileEntityWithId. It will throw and return null.

We are handling only at `FileEntity.fromTransaction`. If we get a error executing the gql request, for instance a connection issue, it will throw an exception that won't be catch on Cubit

```dart
Future<FileEntity?> getLatestFileEntityWithId(String fileId,
      [SecretKey? fileKey]) async {
    final firstOwnerQuery = await _gql.execute(FirstFileEntityWithIdOwnerQuery(
        variables: FirstFileEntityWithIdOwnerArguments(fileId: fileId)));

    if (firstOwnerQuery.data!.transactions.edges.isEmpty) {
      return null;
    }

    final fileOwner =
        firstOwnerQuery.data!.transactions.edges.first.node.owner.address;

    final latestFileQuery = await _gql.execute(LatestFileEntityWithIdQuery(
        variables:
            LatestFileEntityWithIdArguments(fileId: fileId, owner: fileOwner)));

    final queryEdges = latestFileQuery.data!.transactions.edges;
    if (queryEdges.isEmpty) {
      return null;
    }

    final fileTx = queryEdges.first.node;
    final fileDataRes = await client.api.getSandboxedTx(fileTx.id);

    try {
      return await FileEntity.fromTransaction(
        fileTx,
        fileDataRes.bodyBytes,
        fileKey: fileKey,
      );
    } on EntityTransactionParseException catch (parseException) {
      print(
        'Failed to parse transaction '
        'with id ${parseException.transactionId}',
      );
      return null;
    }
  }

**Comment by jdaev** [comment]
That's the case for pretty much every query and network request in the app.

File: lib/blocs/shared_file/shared_file_cubit.dart [hunk 6]
```diff
@@ -1,33 +1,88 @@
...
+
+  final String fileId;
 
   /// The [SecretKey] that can be used to decode the target file.
   ///
   /// `null` if the file is public.
   final SecretKey? fileKey;
 
-  final ArweaveService? _arweave;
+  final ArweaveService _arweave;
```
Comment (matibat) [inline lib/blocs/shared_file/shared_file_cubit.dart:24]: Loved that! 😎

File: lib/pages/shared_file/shared_file_page.dart [hunk 7]
```diff
@@ -36,6 +40,32 @@ class SharedFilePage extends StatelessWidget {
...
+                      child: ReactiveTextField(
+                        formControlName: 'fileKey',
+                        autofocus: true,
+                        obscureText: true,
+                        decoration: InputDecoration(
+                          labelText: appLocalizationsOf(context).fileKey,
+                        ),
+                        validationMessages: (_) =>
+                            kValidationMessages(appLocalizationsOf(context)),
+                        onEditingComplete: () => context
```
Comment (matibat) [inline lib/pages/shared_file/shared_file_page.dart:57]: So I guess this is the component with the issue 🤔

**Review by matibat — APPROVED** [review]
LGTM. I will update the translation tasks based on the ARB file

**Review by karlprieb — APPROVED** [review]

File: lib/pages/shared_file/shared_file_page.dart [hunk 8]
```diff
@@ -36,6 +40,32 @@ class SharedFilePage extends StatelessWidget {
...
+                            .read<SharedFileCubit>()
+                            .form
+                            .updateValueAndValidity(),
+                      ),
+                    ),
+                    const SizedBox(height: 16),
+                    ElevatedButton(
+                      onPressed: () => context.read<SharedFileCubit>().submit(),
+                      child: Text(appLocalizationsOf(context).unlockEmphasized),
```
Comment (matibat) [inline lib/pages/shared_file/shared_file_page.dart:66]: We are using the emphasized version (capitalized text)
